goreleaser/configure.go

```diff
@@ -0,0 +1,223 @@
+//go:build releaser
+
+package main
+
+// This file is a script which generates the .goreleaser.yaml file for all
+// supported OpenTelemetry Collector distributions.
+//
+// Run it with `make generate-goreleaser`.
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/nfpm/v2/files"
+	yaml "gopkg.in/yaml.v2"
+)
+
+var (
+	ImagePrefixes = []string{"otel"}
+	Architectures = []string{"386", "amd64", "arm64"}
+
+	distsFlag = flag.String("d", "", "Collector distributions(s) to build, comma-separated")
+)
+
+func main() {
+	flag.Parse()
+
+	if len(*distsFlag) == 0 {
+		log.Fatal("no distributions to build")
+	}
+	dists := strings.Split(*distsFlag, ",")
+
+	project := Generate(ImagePrefixes, dists)
+
+	if err := yaml.NewEncoder(os.Stdout).Encode(&project); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func Generate(imagePrefixes []string, dists []string) config.Project {
+	return config.Project{
+		ProjectName: "opentelemetry-collector-releases",
+		Checksum: config.Checksum{
+			NameTemplate: "{{ .ProjectName }}_checksums.txt",
+		},
+
+		Builds:          Builds(dists),
+		Archives:        Archives(dists),
+		NFPMs:           Packages(dists),
+		Dockers:         DockerImages(imagePrefixes, dists),
+		DockerManifests: DockerManifests(imagePrefixes, dists),
+	}
+}
+
+func Builds(dists []string) (r []config.Build) {
+	for _, dist := range dists {
+		r = append(r, Build(dist))
+	}
+	return
+}
+
+// Build configures a goreleaser build.
+// https://goreleaser.com/customization/build/
+func Build(dist string) config.Build {
+	return config.Build{
+		ID:      dist,
+		Dir:     path.Join("distributions", dist, "_build"),
+		Binary:  dist,
+		Env:     []string{"CGO_ENABLED=0"},
+		Flags:   []string{"-trimpath"},
+		Ldflags: []string{"-s", "-w"},
+
+		Goos:   []string{"darwin", "linux", "windows"},
+		Goarch: Architectures,
+		Ignore: []config.IgnoredBuild{
+			{Goos: "windows", Goarch: "arm64"},
+		},
+	}
+}
+
+func Archives(dists []string) (r []config.Archive) {
```

Comment:
Looks good for now, but for the future, should we consider having each goreleaser section in its own source file?

Reply:
iirc, `package main` cannot be split into multiple files when using `go run`. But the sections could be in a non-main imported package.

However, unless/until there’s a big increase in complexity in the generation code, I think it’s better this way anyway. At least, speaking for myself as someone who recently had to figure out how `.goreleaser.yml` was generated, I struggled a bit with all the small YAML snippets in different files.

It’d still have been easier to follow in multiple Go files, with top-level construction of `config.Project` being implicit vs. `yq` merging, and the ability to jump to definitions, but at ~200 lines of code I don’t think it’s worth it. Here the steps are laid out in the same order that goreleaser performs them – you can look at the CI build log and skim this file and understand quickly exactly how the setup works.